lms/services/grading_info.py

```diff
@@ -83,33 +73,31 @@ def get_students_for_grading(
 
         return students
 
-    def upsert_from_request(self, request):
+    def upsert(self, lti_user, lis_result_sourcedid, lis_outcome_service_url):
         """
         Update or create a record based on the LTI params found in the request.
 
         This function will do nothing if the correct parameters cannot be
-        found in the request.
-
-        :arg request: A pyramid request
+        found.
         """
-        try:
-            parsed_params = self._ParamsSchema(request).parse()
-        except ValidationError:
-            # We're missing something we need in the request.
+        if not lis_result_sourcedid or not lis_outcome_service_url:
+            # We're missing something we needed
             # This can happen if the user is not a student, or if the needed
             # LIS data is not present on the request.
             return None
 
         grading_info = self._find_or_create(
-            application_instance=request.lti_user.application_instance,
-            user_id=request.lti_user.user_id,
-            context_id=parsed_params["context_id"],
-            resource_link_id=parsed_params["resource_link_id"],
+            application_instance=lti_user.application_instance,
+            user_id=lti_user.user_id,
+            context_id=lti_user.lti.course_id,
```

Comment:
LTI user has this and is available everytime it makes sense, launches, API requests...